### PR TITLE
Support referring to types as variables

### PR DESCRIPTION
**Pull request by segiddins**

### CODE REVIEW

**Review by harlanhaskins — APPROVED** [review]
This looks great!

File: Sources/Semantic Analysis/Sema.swift [hunk 1]
```diff
@@ -475,7 +475,7 @@ class Sema: ASTTransformer, Pass {
     } else if let decl = context.decl(for: DataType(name: expr.name.name)) {
       expr.isTypeVar = true
       expr.decl = decl
-      expr.type = DataType(name: expr.name.name)
+      expr.type = context.stdlib!.mirror.type //?? DataType(name: expr.name.name)
```
Comment (segiddins) [inline Sources/Semantic Analysis/Sema.swift]: handle this being nil

File: Sources/IRGen/BaseIRGen.swift [hunk 2]
```diff
@@ -471,6 +471,17 @@ class IRGenerator: ASTVisitor, Pass {
...
+      let mirrorInitializer = context.stdlib!.mirrorReflectingTypeMetadataInitializer
+      let function = codegenFunctionPrototype(mirrorInitializer)
+      return VarBinding(ref: function,
+                        storage: .reference,
+                        read: {
+                          let meta = self.codegenTypeMetadata(decl.type)
+                          let arg = self.builder.buildBitCast(meta, type: PointerType(pointee: IntType.int8))
+                          return self.builder.buildCall(function, args: [arg])
+                        },
+                        write: {_ in fatalError("Cannot reassign type")})
```
Comment (segiddins) [inline Sources/IRGen/BaseIRGen.swift]: spacing

File: Sources/Semantic Analysis/TypeChecker.swift [hunk 3]
```diff
@@ -175,11 +175,12 @@ class TypeChecker: ASTTransformer, Pass {
   }
   
   override func visitVarExpr(_ expr: VarExpr) -> Result {
+    if expr.isTypeVar { return }
```
Comment (segiddins) [inline Sources/Semantic Analysis/TypeChecker.swift]: this... seems like a hack?

**Pull request merged**